File: src/backend/Domain/Department.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace MoraKnode.Domain;

/// <summary>
/// Functional org unit. One Resource sits in at most one Department
/// (Resource.DepartmentId), forming the "where do you report" axis of the
/// matrix. The "what do you work on" axis is Project.MemberResourceIds.
/// Two axes ⊥ → matrix view = department × project.
///
/// Tree shape: single parent (DAG / multi-parent deferred to v1.1). Cycle
/// prevention enforced application-side in the repository on update —
/// Mongo can't express "no cycles in self-referential graph" cheaply.
/// </summary>
public class Department
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = null!;

    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }

    [BsonRepresentation(BsonType.ObjectId)]
    public string? ParentDepartmentId { get; set; }

    /// <summary>Optional department head. Used for routing approvals
    /// later; left nullable so a freshly created department doesn't need
    /// a lead picked.</summary>
    [BsonRepresentation(BsonType.ObjectId)]
    public string? LeadResourceId { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

File: src/frontend/lib/models/department.dart
/// Mirrors src/backend/Domain/Department.cs. The matrix's "functional"
/// axis: every Resource sits in at most one Department, and Departments
/// form a single-parent tree (multi-parent / DAG deferred to v1.1). A
/// matrix filter against a department subtree resolves to "every
/// resource reporting under this node, transitively".
class Department {
  final String? id;
  final String name;
  final String? description;
  final String? parentDepartmentId;
  final String? leadResourceId;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Department({
    this.id,
    required this.name,
    this.description,
    this.parentDepartmentId,
    this.leadResourceId,
    this.createdAt,
    this.updatedAt,
  });

  factory Department.fromJson(Map<String, dynamic> json) => Department(
        id: json['id'] as String?,
        name: json['name'] as String,
        description: json['description'] as String?,
        parentDepartmentId: json['parentDepartmentId'] as String?,
        leadResourceId: json['leadResourceId'] as String?,
        createdAt: _parseDate(json['createdAt']),
        updatedAt: _parseDate(json['updatedAt']),
      );

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        'name': name,
        if (description != null) 'description': description,
        if (parentDepartmentId != null) 'parentDepartmentId': parentDepartmentId,
        if (leadResourceId != null) 'leadResourceId': leadResourceId,
      };

  static DateTime? _parseDate(Object? v) =>
      v == null ? null : DateTime.parse(v as String).toUtc();
}
